### PR TITLE
C++: Fix broken qhelp links

### PR DESCRIPTION
**Pull request by MathiasVP**
Fixes https://github.com/github/codeql-c-analysis-team/issues/152. I only had to find a new reference for the date-related queries.

### CODE REVIEW

File: cpp/ql/src/Best Practices/UseOfGoto.qhelp [hunk 1]
```diff
@@ -41,7 +41,7 @@ this rule.
   E. W. Dijkstra Archive: <a href="http://www.cs.utexas.edu/users/EWD/transcriptions/EWD02xx/EWD215.html">A Case against the GO TO Statement (EWD-215)</a>.
 </li>
 <li>
-  MSDN Library: <a href="http://msdn.microsoft.com/en-gb/library/b34dt9cd%28v=vs.80%29.aspx">The goto Statement</a>.
+  MSDN Library: <a href="https://docs.microsoft.com/en-us/cpp/cpp/goto-statement-cpp">The goto Statement</a>.
```
Comment (hubwriter) [inline cpp/ql/src/Best Practices/UseOfGoto.qhelp:44]:
```suggestion
  MSDN Library: <a href="https://docs.microsoft.com/en-us/cpp/cpp/goto-statement-cpp">goto Statement (C++)</a>.
```

File: cpp/ql/src/Likely Bugs/Arithmetic/BadCheckOdd.qhelp [hunk 2]
```diff
@@ -23,7 +23,7 @@ As a result, this check incorrectly considers all negative numbers as even.
 <references>
 
 <li>
-  MSDN Library: <a href="http://msdn.microsoft.com/en-us/library/ty2ax9z9%28v=vs.71%29.aspx">Multiplicative Operators: *, /, and %</a>.
+  MSDN Library: <a href="https://docs.microsoft.com/en-us/cpp/cpp/multiplicative-operators-and-the-modulus-operator">Multiplicative Operators: *, /, and %</a>.
```
Comment (hubwriter) [inline cpp/ql/src/Likely Bugs/Arithmetic/BadCheckOdd.qhelp:26]: Match the title on the target page
```suggestion
  MSDN Library: <a href="https://docs.microsoft.com/en-us/cpp/cpp/multiplicative-operators-and-the-modulus-operator">Multiplicative Operators and the Modulus Operator</a>.
```

File: cpp/ql/src/Likely Bugs/Arithmetic/ComparisonPrecedence.qhelp [hunk 3]
```diff
@@ -21,7 +21,7 @@ It is best to fully parenthesize complex comparison expressions to explicitly de
 <references>
 
 <li>
-  <a href="http://msdn.microsoft.com/en-us/library/126fe14k%28v=VS.80%29.aspx">Operator Precedence and Associativity</a>
+  <a href="https://docs.microsoft.com/en-us/cpp/cpp/cpp-built-in-operators-precedence-and-associativity">Operator Precedence and Associativity</a>
```
Comment (hubwriter) [inline cpp/ql/src/Likely Bugs/Arithmetic/ComparisonPrecedence.qhelp:24]: C++ built-in operators, precedence, and associativity
```suggestion
  MSDN Library: <a href="https://docs.microsoft.com/en-us/cpp/cpp/cpp-built-in-operators-precedence-and-associativity">C++ built-in operators, precedence, and associativity</a>
```

File: cpp/ql/src/Likely Bugs/Arithmetic/IntMultToLong.qhelp [hunk 4]
```diff
@@ -23,7 +23,7 @@ the expression would produce a result that would be too large to fit in the smal
 <references>
 
 <li>
-  MSDN Library: <a href="http://msdn.microsoft.com/en-us/library/ty2ax9z9%28v=vs.71%29.aspx">Multiplicative Operators: *, /, and %</a>.
+  MSDN Library: <a href="https://docs.microsoft.com/en-us/cpp/cpp/multiplicative-operators-and-the-modulus-operator">Multiplicative Operators: *, /, and %</a>.
```
Comment (hubwriter) [inline cpp/ql/src/Likely Bugs/Arithmetic/IntMultToLong.qhelp:26]:
```suggestion
  MSDN Library: <a href="https://docs.microsoft.com/en-us/cpp/cpp/multiplicative-operators-and-the-modulus-operator">Multiplicative Operators and the Modulus Operator</a>.
```

File: cpp/ql/src/Likely Bugs/Format/WrongTypeFormatArguments.qhelp [hunk 5]
```diff
@@ -23,7 +23,7 @@ the function.
 <li>CERT C Coding
 Standard: <a href="https://www.securecoding.cert.org/confluence/display/c/FIO30-C.+Exclude+user+input+from+format+strings">FIO30-C. Exclude user input from format strings</a>.</li>
 <li>cplusplus.com: <a href="http://www.tutorialspoint.com/cplusplus/cpp_functions.htm">C++ Functions</a>.</li>
-<li>MSDN Alphabetical Function Reference: <a href="http://msdn.microsoft.com/en-us/library/wc7014hz%28VS.71%29.aspx">printf, wprintf</a>.</li>
+<li>MSDN Alphabetical Function Reference: <a href="https://docs.microsoft.com/en-us/cpp/c-runtime-library/reference/printf-printf-l-wprintf-wprintf-l">printf, wprintf</a>.</li>
```
Comment (hubwriter) [inline cpp/ql/src/Likely Bugs/Format/WrongTypeFormatArguments.qhelp:26]:
```suggestion
<li>MSDN Alphabetical Function Reference: <a href="https://docs.microsoft.com/en-us/cpp/c-runtime-library/reference/printf-printf-l-wprintf-wprintf-l">printf, _printf_l, wprintf, _wprintf_l</a>.</li>
```

File: cpp/ql/src/Likely Bugs/Likely Typos/MissingEnumCaseInSwitch.qhelp [hunk 6]
```diff
@@ -23,7 +23,7 @@ indication that there may be cases unhandled by the <code>switch</code> statemen
   Tutorialspoint - The C++ Programming Language: <a href="http://www.tutorialspoint.com/cplusplus/cpp_switch_statement.htm">C++ switch statement</a>
 </li>
 <li>
-  MSDN Library: <a href="http://msdn.microsoft.com/en-us/library/k0t5wee3%28v=VS.80%29.aspx">The switch Statement</a>
+  MSDN Library: <a href="https://docs.microsoft.com/en-us/cpp/cpp/switch-statement-cpp">The switch Statement</a>
```
Comment (hubwriter) [inline cpp/ql/src/Likely Bugs/Likely Typos/MissingEnumCaseInSwitch.qhelp:26]:
```suggestion
  MSDN Library: <a href="https://docs.microsoft.com/en-us/cpp/cpp/switch-statement-cpp">switch statement (C++)</a>
```

File: cpp/ql/src/Likely Bugs/Memory Management/SuspiciousCallToMemset.qhelp [hunk 7]
```diff
@@ -30,7 +30,7 @@ For an array, the size is the number of elements of the array multiplied by the
   Cplusplus.comn: <a href="http://www.cplusplus.com/reference/clibrary/cstring/memset/">memset</a>
 </li>
 <li>
-  MSDN Library: <a href="http://msdn.microsoft.com/en-us/library/aa246471%28v=VS.60%29.aspx">memset</a>, <a href="http://msdn.microsoft.com/en-us/library/4s7x1k91%28v=VS.71%29.aspx">sizeof Operator</a>
+  MSDN Library: <a href="https://docs.microsoft.com/en-us/cpp/c-runtime-library/reference/memset-wmemset">memset</a>, <a href="https://docs.microsoft.com/en-us/cpp/cpp/sizeof-operator">sizeof Operator</a>
```
Comment (hubwriter) [inline cpp/ql/src/Likely Bugs/Memory Management/SuspiciousCallToMemset.qhelp:33]:
```suggestion
  MSDN Library: <a href="https://docs.microsoft.com/en-us/cpp/c-runtime-library/reference/memset-wmemset">memset, wmemset</a>, <a href="https://docs.microsoft.com/en-us/cpp/cpp/sizeof-operator">sizeof Operator</a>
```

File: cpp/ql/src/Metrics/Files/FMacroRatio.qhelp [hunk 8]
```diff
@@ -27,7 +27,7 @@ and IDE support than macros.</p>
 <references>
 
 <li>
-  <a href="http://msdn.microsoft.com/en-us/library/503x3e3s%28v=vs.80%29.aspx">Macros</a>
+  <a href="https://docs.microsoft.com/en-us/cpp/preprocessor/macros-c-cpp">Macros</a>
```
Comment (hubwriter) [inline cpp/ql/src/Metrics/Files/FMacroRatio.qhelp:30]:
```suggestion
  MSDN Library: <a href="https://docs.microsoft.com/en-us/cpp/preprocessor/macros-c-cpp">Macros (C/C++)</a>
```

File: cpp/ql/src/Metrics/Files/FTimeInFrontend.qhelp [hunk 9]
```diff
@@ -15,7 +15,7 @@
 <references>
 
 <li>
-  <a href="http://msdn.microsoft.com/en-us/library/36k2cdd4%28v=VS.80%29.aspx">The #include Directive</a>
+  <a href="https://docs.microsoft.com/en-us/cpp/preprocessor/hash-include-directive-c-cpp">The #include Directive</a>
```
Comment (hubwriter) [inline cpp/ql/src/Metrics/Files/FTimeInFrontend.qhelp:18]:
```suggestion
  MSDN Library: <a href="https://docs.microsoft.com/en-us/cpp/preprocessor/hash-include-directive-c-cpp">#include directive (C/C++)</a>
```

File: cpp/ql/src/Security/CWE/CWE-089/SqlTainted.qhelp [hunk 10]
```diff
@@ -21,7 +21,7 @@ string before converting it to SQL.</p>
 </example>
 <references>
 
-<li>Microsoft Developer Network: <a href="http://msdn.microsoft.com/en-us/library/ms161953.aspx">SQL Injection</a>.</li>
+<li>Microsoft Developer Network: <a href="https://docs.microsoft.com/en-us/sql/relational-databases/security/sql-injection">SQL Injection</a>.</li>
```
Comment (hubwriter) [inline cpp/ql/src/Security/CWE/CWE-089/SqlTainted.qhelp:24]:
```suggestion
<li>MSDN Library: <a href="https://docs.microsoft.com/en-us/sql/relational-databases/security/sql-injection">SQL Injection</a>.</li>
```

File: cpp/ql/src/jsf/4.10 Classes/AV Rule 89.qhelp [hunk 11]
```diff
@@ -15,7 +15,7 @@ as both virtual and non-virtual in one hierarchy as violations.</p>
   AV Rule 89, <em>Joint Strike Fighter Air Vehicle C++ Coding Standards</em>. Lockheed Martin Corporation, 2005.
 </li>
 <li>
-  <a href="http://msdn.microsoft.com/en-us/library/wcz57btd%28v=vs.80%29.aspx">Virtual Base Classes</a>
+  <a href="https://docs.microsoft.com/en-us/cpp/cpp/multiple-base-classes">Virtual Base Classes</a>
```
Comment (hubwriter) [inline cpp/ql/src/jsf/4.10 Classes/AV Rule 89.qhelp:18]:
```suggestion
  MSDN Library: <a href="https://docs.microsoft.com/en-us/cpp/cpp/multiple-base-classes">Multiple Base Classes</a>.
```

File: cpp/ql/src/jsf/4.13 Functions/AV Rule 114.qhelp [hunk 12]
```diff
@@ -37,7 +37,7 @@ assuming the function had a non-float return type that can fit in a machine word
   MISRA C++ Rule 8-4-3, <em>Guidelines for the use of the C++ language in critical systems</em>. The Motor Industry Software Reliability Associate, 2008.
 </li>
 <li>
-  <a href="http://msdn.microsoft.com/en-US/library/k68ktdwf%28v=VS.80%29.aspx">The return Statement</a>
+  <a href="https://docs.microsoft.com/en-us/cpp/cpp/return-statement-cpp">The return Statement</a>
```
Comment (hubwriter) [inline cpp/ql/src/jsf/4.13 Functions/AV Rule 114.qhelp:40]:
```suggestion
  MSDN Library: <a href="https://docs.microsoft.com/en-us/cpp/cpp/return-statement-cpp">return Statement (C++)</a>
```

File: cpp/ql/src/jsf/4.16 Initialization/AV Rule 145.qhelp [hunk 13]
```diff
@@ -37,7 +37,7 @@ better to let the compiler assign values at compile time.
   MISRA C++ Rule 8-5-3, <em>Guidelines for the use of the C++ language in critical systems</em>. The Motor Industry Software Reliability Associate, 2008.
 </li>
 <li>
-  <a href="http://msdn.microsoft.com/en-us/library/2dzy4k6e%28v=vs.71%29.aspx">C++ Enumeration Declarations</a>
+  <a href="https://docs.microsoft.com/en-us/cpp/cpp/enumerations-cpp">C++ Enumeration Declarations</a>
```
Comment (hubwriter) [inline cpp/ql/src/jsf/4.16 Initialization/AV Rule 145.qhelp:40]:
```suggestion
  MSDN Library: <a href="https://docs.microsoft.com/en-us/cpp/cpp/enumerations-cpp">Enumerations (C++)</a>.
```

File: cpp/ql/src/jsf/4.24 Control Flow Structures/AV Rule 189.qhelp [hunk 14]
```diff
@@ -29,7 +29,7 @@ the break statement only exits from one level of the loop.</p>
   <a href="http://www.cs.utexas.edu/users/EWD/transcriptions/EWD02xx/EWD215.html">A Case against the GO TO Statement (EWD-215).</a>
 </li>
 <li>
-  <a href="https://msdn.microsoft.com/en-gb/library/b34dt9cd%28v=vs.80%29.aspx">MSDN Library: The goto Statement</a>
+  <a href="https://docs.microsoft.com/en-us/cpp/cpp/goto-statement-cpp">MSDN Library: The goto Statement</a>
```
Comment (hubwriter) [inline cpp/ql/src/jsf/4.24 Control Flow Structures/AV Rule 189.qhelp:32]:
```suggestion
  MSDN Library: <a href="https://docs.microsoft.com/en-us/cpp/cpp/goto-statement-cpp">goto Statement (C++)</a>.
```

**Review by hubwriter — COMMENTED** [review]
I've added suggested changes to use the titles that are used on the references pages. Otherwise this LGTM.

File: cpp/ql/src/Likely Bugs/Format/WrongTypeFormatArguments.qhelp [hunk 15]
```diff
@@ -23,7 +23,7 @@ the function.
 <li>CERT C Coding
 Standard: <a href="https://www.securecoding.cert.org/confluence/display/c/FIO30-C.+Exclude+user+input+from+format+strings">FIO30-C. Exclude user input from format strings</a>.</li>
 <li>cplusplus.com: <a href="http://www.tutorialspoint.com/cplusplus/cpp_functions.htm">C++ Functions</a>.</li>
-<li>MSDN Alphabetical Function Reference: <a href="http://msdn.microsoft.com/en-us/library/wc7014hz%28VS.71%29.aspx">printf, wprintf</a>.</li>
+<li>MSDN Alphabetical Function Reference: <a href="https://docs.microsoft.com/en-us/cpp/c-runtime-library/reference/printf-printf-l-wprintf-wprintf-l">printf, _printf_l, wprintf, _wprintf_l</a>.</li>
```
Comment (geoffw0) [inline cpp/ql/src/Likely Bugs/Format/WrongTypeFormatArguments.qhelp:26]:
```suggestion
<li>CRT Alphabetical Function Reference: <a href="https://docs.microsoft.com/en-us/cpp/c-runtime-library/reference/printf-printf-l-wprintf-wprintf-l">printf, _printf_l, wprintf, _wprintf_l</a>.</li>
```
Again they appear to have rebranded.

File: cpp/ql/src/Likely Bugs/Leap Year/Adding365DaysPerYear.qhelp [hunk 16]
```diff
@@ -15,7 +15,7 @@ of days.  Alternatively, use an established library routine that already contain
 </recommendation>
 
 <references>
-  <li>U.S. Naval Observatory Website - <a href="https://aa.usno.navy.mil/faq/docs/calendars.php"> Introduction to Calendars</a></li>
+  <li>NASA's Goddard Space Flight Center Website - <a href="https://eclipse.gsfc.nasa.gov/SEhelp/calendars.html"> Calendars</a></li>
```
Comment (geoffw0) [inline cpp/ql/src/Likely Bugs/Leap Year/Adding365DaysPerYear.qhelp:18]:
```suggestion
  <li>NASA / Goddard Space Flight Center - <a href="https://eclipse.gsfc.nasa.gov/SEhelp/calendars.html">Calendars</a></li>
```
This is admittedly nitpicky, but I think it's better to match the page's title.

File: cpp/ql/src/Likely Bugs/Leap Year/UncheckedLeapYearAfterYearModification.qhelp [hunk 17]
```diff
@@ -22,7 +22,7 @@
 </example>
 
 <references>
-  <li>U.S. Naval Observatory Website - <a href="https://aa.usno.navy.mil/faq/docs/calendars.php"> Introduction to Calendars</a></li>
+  <li>NASA's Goddard Space Flight Center Website - <a href="https://eclipse.gsfc.nasa.gov/SEhelp/calendars.html"> Calendars</a></li>
```
Comment (geoffw0) [inline cpp/ql/src/Likely Bugs/Leap Year/UncheckedLeapYearAfterYearModification.qhelp:25]:
```suggestion
  <li>NASA / Goddard Space Flight Center - <a href="https://eclipse.gsfc.nasa.gov/SEhelp/calendars.html">Calendars</a></li>
```

File: cpp/ql/src/Likely Bugs/Leap Year/UncheckedReturnValueForTimeFunctions.qhelp [hunk 18]
```diff
@@ -34,7 +34,7 @@
 </example>
 
 <references>
-  <li>U.S. Naval Observatory Website - <a href="https://aa.usno.navy.mil/faq/docs/calendars.php"> Introduction to Calendars</a></li>
+  <li>NASA's Goddard Space Flight Center Website - <a href="https://eclipse.gsfc.nasa.gov/SEhelp/calendars.html"> Calendars</a></li>
```
Comment (geoffw0) [inline cpp/ql/src/Likely Bugs/Leap Year/UncheckedReturnValueForTimeFunctions.qhelp:37]:
```suggestion
  <li>NASA / Goddard Space Flight Center - <a href="https://eclipse.gsfc.nasa.gov/SEhelp/calendars.html">Calendars</a></li>
```

File: cpp/ql/src/Likely Bugs/Leap Year/UnsafeArrayForDaysOfYear.qhelp [hunk 19]
```diff
@@ -23,7 +23,7 @@
 </example>
 
 <references>
-  <li>U.S. Naval Observatory Website - <a href="https://aa.usno.navy.mil/faq/docs/calendars.php"> Introduction to Calendars</a></li>
+  <li>NASA's Goddard Space Flight Center Website - <a href="https://eclipse.gsfc.nasa.gov/SEhelp/calendars.html"> Calendars</a></li>
```
Comment (geoffw0) [inline cpp/ql/src/Likely Bugs/Leap Year/UnsafeArrayForDaysOfYear.qhelp:26]:
```suggestion
  <li>NASA / Goddard Space Flight Center - <a href="https://eclipse.gsfc.nasa.gov/SEhelp/calendars.html">Calendars</a></li>
```

File: cpp/ql/src/Likely Bugs/NestedLoopSameVar.qhelp [hunk 20]
```diff
@@ -26,7 +26,7 @@ outer loop. </p>
   Tutorialspoint - The C++ Programming Language: <a href="http://www.tutorialspoint.com/cplusplus/cpp_nested_loops.htm">C++ nested loops</a>
 </li>
 <li>
-  MSDN Library: <a href="http://msdn.microsoft.com/en-us/library/8y82wx12%28v=VS.80%29.aspx">Nested Control Structures</a>
+  MSDN Library: <a href="https://docs.microsoft.com/en-us/dotnet/visual-basic/programming-guide/language-features/control-flow/nested-control-structures">Nested Control Structures</a>
```
Comment (geoffw0) [inline cpp/ql/src/Likely Bugs/NestedLoopSameVar.qhelp:29]: Strange that this page is about Visual Basic, in the C/C++ query help.  The points on the page apply perfectly well to C/C++, but I can't help but think we could do better.

Comment (geoffw0) [reply cpp/ql/src/Likely Bugs/NestedLoopSameVar.qhelp:29]: Perhaps https://www.programiz.com/cpp-programming/nested-loops ?

**Review by geoffw0 — COMMENTED** [review]
A few nitpicks, otherwise LGTM, it's good to maintain working links!

File: cpp/ql/src/Likely Bugs/Arithmetic/ComparisonPrecedence.qhelp [hunk 21]
```diff
@@ -21,7 +21,7 @@ It is best to fully parenthesize complex comparison expressions to explicitly de
 <references>
 
 <li>
-  <a href="http://msdn.microsoft.com/en-us/library/126fe14k%28v=VS.80%29.aspx">Operator Precedence and Associativity</a>
+  MSDN Library: <a href="https://docs.microsoft.com/en-us/cpp/cpp/cpp-built-in-operators-precedence-and-associativity">C++ built-in operators, precedence, and associativity</a>
```
[the remark this answers is not in the document]
Comment (MathiasVP) [reply cpp/ql/src/Likely Bugs/Arithmetic/ComparisonPrecedence.qhelp:24]: Good point. I think it's a good idea, but maybe it should be discussed in a separate issue.

File: cpp/ql/src/Likely Bugs/NestedLoopSameVar.qhelp (hunk 20 above)
Comment (MathiasVP) [reply cpp/ql/src/Likely Bugs/NestedLoopSameVar.qhelp:29]: I agree that it's a strange choice. I looked on [Wayback Machine](https://web.archive.org/web/20110407080413/http://msdn.microsoft.com/en-us/library/8y82wx12(v=VS.80).aspx) at the old link, and that apparently also referenced Visual Basic. If we don't mind linking to tutorials then I think your link explains the concepts equally well (and with the bonus of being written in C++!)

Comment (MathiasVP) [reply cpp/ql/src/Likely Bugs/NestedLoopSameVar.qhelp:29]: Actually... after looking at the qhelp file again I realized that we already have this one linked:
```html
Tutorialspoint - The C++ Programming Language: <a href="http://www.tutorialspoint.com/cplusplus/cpp_nested_loops.htm">C++ nested loops</a>
```
so maybe adding https://www.programiz.com/cpp-programming/nested-loops is somewhat redundant. What if we just remove that link altogether?

Comment (geoffw0) [reply cpp/ql/src/Likely Bugs/NestedLoopSameVar.qhelp:29]: I'm happy with it just removed.

Comment (MathiasVP) [reply cpp/ql/src/Likely Bugs/NestedLoopSameVar.qhelp]: Removed in 9db66a1b9451c08122eabca386d0b1b40f317021.

**Review by geoffw0 — APPROVED** [review]
LGTM, and has had multiple reviews now.

**Pull request merged**